feat: fuzon-http implementation

cmdoret opened the pull request:
Adds a webserver to deploy fuzon as a microservice.
It has the following features:
* load key: ontology pairs (aka collections) from a config file
* cli to provide config file as parameter
* /top endpoint to retrieve top N codes from a collection matching input text
* /list endpoint to retrieve list of collections

supermaxiste commented on fuzon-http/README.md:
```diff
@@ -0,0 +1,23 @@
+
+This is a web-server to deploy fuzon as a web-service.
+All ontologies are loaded once on server startup, and the indices are kept in memory.
+
```
**suggestion**: around the top of the README I would add the gif you shared with us, maybe with a focus on the tool showing the top X matching terms. That's an easy way to see the tool in action and understand what it does. Alternatively the gif can also go to the main README to showcase it even more.

cmdoret replied:
Good idea! I've added a minimal (15 lines of bash) client example to the readme in f5b8595 to show how to interact with the server, and a gif of what it may look like.

supermaxiste commented on fuzon-http/README.md:
````diff
@@ -0,0 +1,23 @@
+```shell
+cd fuzon-http
+cargo build --release
+../target/release/fuzon-http --config config/example.json
````
**issue(minor, nitpick)**: either we rename the section "installation and run" or we move L15 to a separate code chunk and define it "run"

cmdoret replied:
:+1: fixed in 2c79e4f

supermaxiste commented on fuzon-http/README.md:
````diff
@@ -0,0 +1,23 @@
+../target/release/fuzon-http --config config/example.json
+```
+
+## Usage
````
**issue (non-blocking)**: this usage section is not very helpful to a newcomer. After running the installation above, the server runs but it's not clear how a user is supposed to interact with it. I suggest we go a bit more in detail because we're the ones setting (for example) the port in the code, but that's not clear here. 
Additionally, for someone less experienced in APIs, showing only the commands doesn't tell much. Some examples on what the output would look like and how the command `/top?collection={collection}&top={top}&query={query}`  works would be helpful. We can use the example config as an example 👍

cmdoret replied:
Indeed, I've added the host+port to the config file!

cmdoret replied:
added examples based on config file

supermaxiste commented on fuzon-http/src/main.rs:
```diff
@@ -37,16 +40,13 @@ impl AppState {
             .collections
             .into_iter()
-            .map(|(k, v)| (k, TermMatcher::from_paths(vec![&v]).unwrap())).collect();
+            .inspect(|(k, _)| println!("Loading {}...", k))
```
**suggestion(non-blocking, nitpick)**: since we call them "collections" I would also print "Loading collection {}..."
```suggestion
            .inspect(|(k, _)| println!("Loading collection {}...", k))
```

supermaxiste commented on fuzon-http/src/main.rs:
```diff
@@ -75,28 +75,34 @@ async fn top(data: web::Data<AppState>, req: web::Query<CodeRequest>) -> Result<
-        }"#;
     let data = web::block(move || 
-        serde_json::from_str::<Config>(config_data).unwrap()
+        AppState::from_config(config)
```
**note(non-blocking)**: I would need an explanation (in person) on why this solution suddently fixes the concurrency problems. Good job finding it out ;)

cmdoret replied:
I think it works because of `web::block`. I was not able to use it before, because it requires cloning the instance of `AppState`, along with its internal `TermMatcher`s and they did not implement `Clone`, after deriving Clone on everything, this solution worked :p

supermaxiste reviewed (changes requested):
**praise**: code looks clean. I tried to look for small improvements, but the code is simple and very self explanatory. Feels very Groovy somehow ;)
**praise**: installation works neatly and tool also runs smoothly 🥳 

## Major

**issue**: we need to be more transparent with the user about what is going on in the backend.

1. A message stating which ports are exposed is a good start. A full link on how to access the open port is also very helpful, this way users know where things are happening and what is running.

2. A more descriptive Usage section would be best because APIs are hard, especially when there's not a lot of documentation. Breaking down commands and giving examples is always a plus.

3. We build `fuzon` with `modos` in mind as use case. How can a dev with a tool similar to `modos` set up and interact with `fuzon` via `fuzon-http`? I feel like having an example use case somewhere in the README would be nice. I'm happy to hear your thoughts

## Minor

**thought(non-blocking)**: we're currently hardcoding the port, should that be also part of the config somehow? As mentioned before, as long as we're transparent with the user, we can leave it as is for now.

**suggestion(non-blocking)**: I know the code is not too long, but some unit tests for some of the functions would be great. It can be a separate issue as well.

cmdoret commented:
I've addressed your comments except for:

> A full link on how to access the open port is also very helpful
I could harcode messages into the code, but I think this would better be done through something like an openapi spec, which can be generated using rust packages (e.g. paperclip). But this is probably too big for this PR.

> some unit tests for some of the functions would be great
Simply because most functions require the server to be online for testing and I'm not sure how to do it cleanly, there is prob a way to do this, but I'd rather keep it for another PR.

supermaxiste reviewed (approved):
Very good changes, re-tested everything successfully and all works like a charm 👏 

**praise**: the future-proof API structure is an excellent point I'll keep in mind
**praise**: README is great and the gif is great ⭐ 

For the unit tests & the API spec I'll open separate enhancement issues 👍

cmdoret commented:
Thanks @supermaxiste indeed, let's tackle #22  and #21  separately :+1:
